File: javascript/heap.js
var swap = require('./util').swap

// heap implements heap sort algorithm. The function takes an array.
function heap(a) {
	var n = a.length
	a.unshift(0)
	for(var i = n/2; i >= 1; i--)
		sink(a, i, n)
	for(var i = 1; i <= n; i++) {
		swap(a, 1, n-i+1)
		sink(a, 1, n-i)
	}
	a.shift()
}

function sink(a, i, n) {
	var lc = 2*i
	if(lc > n)
		return
	var rc = lc+1
	var mc = (rc > n) ? lc : (a[lc] > a[rc] ? lc : rc)
	if(a[i] >= a[mc])
		return
	swap(a, i, mc)
	sink(a, mc, n)
}

module.exports = heap

File: java/Heap.java
public class Heap {
	public static void sort(int a[]) {
		int h[] = new int[a.length+1];
		for(int i = 0; i < a.length; i++) {
			h[i+1] = a[i];
		}
		for(int i = a.length/2; i >= 1; i--)
			sink(h, i, a.length);
		for(int i = 1; i <= a.length; i++) {
			Util.swap(h, 1, a.length-i+1);
			sink(h, 1, a.length-i);
		}
		for(int i = 0; i < a.length; i++) {
			a[i] = h[i+1];
		}
	}

	static void sink(int a[], int i, int n) {
		int lc = 2*i;
		if(lc > n)
			return;
		int rc = lc+1;
		int mc = (rc > n) ? lc : (a[lc] > a[rc] ? lc : rc);
		if(a[i] >= a[mc])
			return;
		Util.swap(a, i, mc);
		sink(a, mc, n);
	}
}
